XiEditor/EditViewController.swift

```diff
@@ -850,23 +840,65 @@ class EditViewController: NSViewController, EditViewDataSource, FindDelegate, Sc
     }
     
     public func languageChanged(_ languageIdentifier: String) {
-        self.currentLanguage = languageIdentifier
+        currentLanguage = languageIdentifier
+        settingsStore.saveRecentlyUsed(languages: [languageIdentifier])
+        reloadLanguagesMenu()
     }
     
     public func availableLanguagesChanged(_ languages: [String]) {
-        let languagesMenu = NSApplication.shared.mainMenu!.item(withTitle: "View")!.submenu!.item(withTitle: "Language")!.submenu!;
-        
-        let currentlyActive = languagesMenu.items
-            .filter { $0.state.rawValue == 1 }
-            .first?.title
-        
-        languagesMenu.removeAllItems()
-        for language in languages {
-            let item = NSMenuItem(title: language, action: #selector(EditViewController.debugSetLanguage(_:)),
-                keyEquivalent: "")
-            item.state = NSControl.StateValue(rawValue: (language == currentlyActive) ? 1 : 0)
-            languagesMenu.addItem(item)
+        availableLanguages = languages
+        reloadLanguagesMenu()
+    }
+
+    private func reloadLanguagesMenu() {
+        guard let menu = languagesMenu() else {
+            return
         }
+
+        menu.removeAllItems()
+
+        let recentlyUsedlanguages = settingsStore.fetchRecentlyUsedLanguages()
+        if recentlyUsedlanguages.isEmpty == false {
+            let recentlyUsedItems = buildRecentlyUsedLanguagesSection(from: recentlyUsedlanguages)
+            recentlyUsedItems.forEach(menu.addItem)
+            menu.addItem(NSMenuItem.separator())
+        }
+
+        let availableItems = buildAvailableLanguagesSection(from: self.availableLanguages, currentLanguage: self.currentLanguage ?? "")
+
+        availableItems.forEach(menu.addItem)
+    }
+
+    private func languagesMenu() -> NSMenu? {
+        let mainMenu = NSApplication.shared.mainMenu
+        let viewMenu = mainMenu?.item(withTitle: "View")?.submenu
+        return viewMenu?.item(withTitle: "Language")?.submenu
+    }
+
+    private func buildRecentlyUsedLanguagesSection(from languages: [String]) -> [NSMenuItem] {
+        var items: [NSMenuItem] = []
+        let titleItem = NSMenuItem(title: NSLocalizedString("Recently Used", comment: "Recently Used"), action: nil, keyEquivalent: "")
```

Comment:
Might want to clarify the comment with something like "[Language]"

Reply:
@jansol I did not get your comment. Could you please explain?

Reply:
The comment is meant to provide context to translators so they don't need to look at (and understand!) the code when writing translations. (They should only need to work with the strings file)

Actually on second thought, "Language list subtitle" would be an even better description.